STEP 1:

In [16]:
# Downloading the exploration data for the regions

import pandas as pd
geo_data_0=pd.read_csv('/datasets/geo_data_0.csv')
geo_data_1=pd.read_csv('/datasets/geo_data_1.csv')
geo_data_2=pd.read_csv('/datasets/geo_data_2.csv')

In [17]:
# Preparing the data

# Examining the first 10 rows of dataframe geo_data_0
print('Examining the first 10 rows of data in the geo_data_0:')
display(geo_data_0.head(5))

# Examining the size of Dataframe
print('geo_data_0 size:',geo_data_0.shape)
print()

# Checking 10 random rows to get an idea about the data
print()
print('Examining 10 random rows:')
display(geo_data_0.sample(10))
print()

# Checking for null values
print(geo_data_0.isnull().sum())
print()
print('-'*60)

# Examining the first 10 rows of dataframe geo_data_1
print()
print('Examining the first 10 rows of data in the geo_data_1:')
display(geo_data_1.head(5))

# Examining the size of Dataframe
print('geo_data_1 size:',geo_data_1.shape)
print()

# Checking 10 random rows to get an idea about the data
print()
print('Examining 10 random rows:')
display(geo_data_1.sample(10))
print()

# Checking for null values
print(geo_data_1.isnull().sum())
print()
print('-'*60)

# Examining the first 10 rows of dataframe geo_data_2
print()
print('Examining the first 10 rows of data in the geo_data_2:')
display(geo_data_2.head(5))

# Examining the size of Dataframe
print('geo_data_2 size:',geo_data_2.shape)

# Checking 10 random rows to get an idea about the data
print()
print('Examining 10 random rows:')
display(geo_data_2.sample(10))
print()

# Checking for null values
print(geo_data_2.isnull().sum())
print()
print('-'*60)


Examining the first 10 rows of data in the geo_data_0:


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

geo_data_0 size: (100000, 5)


Examining 10 random rows:


id        f0        f1        f2     product
21943  cfVuD  0.905789 -0.631687  7.303184  137.318759
51341  5XRCB  2.129358  0.338737  2.832552  139.328886
19001  84tzi  0.905971  0.431205  8.662604  148.688475
67185  L2yJv  0.496695  0.754861  4.555801  114.739781
14034  dUHBD  1.866520  0.325221  4.928633  162.813321
21930  jlqJ1  0.201945 -0.064500  1.668350   75.283471
97634  c31eH -0.834594 -0.100108  8.624480   72.801228
33318  RIO1k  0.294333  1.011525  2.655589  120.474481
40072  7nySv  2.035941  0.343006  3.091865  166.183128
73471  Hpt0E  1.254347 -0.742699  4.470766   95.765238


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

------------------------------------------------------------

Examining the first 10 rows of data in the geo_data_1:


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

geo_data_1 size: (100000, 5)


Examining 10 random rows:


id         f0         f1        f2    product
74988  qVHt0   7.088394  -3.577313 -0.000988   0.000000
66434  6vf7r   9.713767   1.472907  2.997605  80.859783
65542  Fpeve  15.587615  -7.364047  1.995681  53.906522
39646  zwuQZ  -5.304472 -10.450874  3.006560  84.038886
84501  X1U0i  -2.161034  -7.548142  3.007392  84.038886
70036  7GisP  16.589990  -1.302135 -0.001723   0.000000
4022   iRoq5   5.134398 -10.522966  1.003737  26.953261
27849  qTkOz  -7.571780  -1.295584  0.007408   3.179103
66227  Domil  -3.498473  -8.750826  0.001475   3.179103
39298  ZK9kl   7.312352  -2.132401  3.007985  80.859783


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

------------------------------------------------------------

Examining the first 10 rows of data in the geo_data_2:


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

geo_data_2 size: (100000, 5)

Examining 10 random rows:


id        f0        f1        f2     product
87468  xG3uE  0.181797  0.913376  5.827309   59.953913
23217  Zzgkv  0.509912 -1.468430  3.362742  133.043525
65501  kBkjO  0.442495 -0.951890 -3.587644   40.104142
30670  0AOkX  0.659698  1.737941  4.540472   62.490989
85050  XPdnM  0.899024  2.342144  5.433823  156.624435
36411  8Jjum -2.082027 -1.118640  7.346449  156.239282
21200  aWkDI  1.138302 -1.242490  3.770624  153.594066
59541  dYg7K  0.053770  0.063273  5.241263  159.828323
29423  mVhRp  0.106826 -3.389697  1.402649   82.815270
16043  12qyr -0.090856 -1.007429 -0.380501   17.560463


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

------------------------------------------------------------


STEP 2:

In [18]:
# Training and testing the model for each region
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Assigning Features and Target for geo_data_0
features_0 = geo_data_0[['f0','f1','f2']]
target_0 = geo_data_0['product']
random_state=11
# Splitting the data into training and validation features for geo_data_0
features_train_0,features_valid_0,target_train_0,target_valid_0=train_test_split(features_0, target_0, test_size=0.25, random_state=random_state)

# Assigning Features and Target for geo_data_1
features_1 = geo_data_1[['f0','f1','f2']]
target_1 = geo_data_1['product']
# Splitting the data into training and validation features for geo_data_1
features_train_1,features_valid_1,target_train_1,target_valid_1=train_test_split(features_1, target_1, test_size=0.25, random_state=random_state)

# Assigning Features and Target for geo_data_2
features_2 = geo_data_2[['f0','f1','f2']]
target_2 = geo_data_2['product']
# Splitting the data into training and validation features for geo_data_2
features_train_2,features_valid_2,target_train_2,target_valid_2=train_test_split(features_2, target_2, test_size=0.25, random_state=random_state)

# Training Model for geo_data_0
model_0 = LinearRegression()
model_0.fit(features_train_0, target_train_0)
predictions_valid_0=model_0.predict(features_valid_0)
result_0=mean_squared_error(target_valid_0,predictions_valid_0)**0.50
print(f"Average predicted reserves for 'Region 0': {predictions_valid_0.mean():.2f}")
print(f"RMSE of the Linear Regression Model on validation set of 'geo_data_0': {result_0:.2f}")
print()

# Training model for geo_data_1
model_1 = LinearRegression()
model_1.fit(features_train_1, target_train_1)
predictions_valid_1= model_1.predict(features_valid_1)
result_1= mean_squared_error(target_valid_1, predictions_valid_1)**0.50
print(f"Average predicted reserves for 'Region 1': {predictions_valid_1.mean():.2f}")
print(f"RMSE of the Linear Regression Model on validation set of 'Region 1': {result_1:.2f}")
print()

# Training m,odel for geo_data_2
model_2= LinearRegression()
model_2.fit(features_train_2, target_train_2)
predictions_valid_2= model_2.predict(features_valid_2)
result_2= mean_squared_error(target_valid_2, predictions_valid_2)**0.50
print(f"Average predicted reserves for 'Region 2': {predictions_valid_2.mean():.2f}")
print(f"RMSE of the Linear Regression Model on validation set of 'geo_data_2': {result_2:.2f}")
print()


Average predicted reserves for 'Region 0': 92.40
RMSE of the Linear Regression Model on validation set of 'geo_data_0': 37.82

Average predicted reserves for 'Region 1': 68.74
RMSE of the Linear Regression Model on validation set of 'Region 1': 0.89

Average predicted reserves for 'Region 2': 95.02
RMSE of the Linear Regression Model on validation set of 'geo_data_2': 40.00



Analysis:
Region 1 has most reliable predictions with RMSE = 0.89 but has lowest reserves: 68.74
Region 0 has moderate reliability with RMSE= 37.82 but high reserves: 92.40
Region 2 has least reliable predictions with RMSE= 40 but highest reserves: 95.02

This tells us that for Region 1, model is very confident in it's predictions but the wells don't contain much oil. Regions 0 and 2 have higher potential reserves but the model is less certain anout the predictions.

As of now, the "best" region can't be determined because it's a trade-off between reliability and higher potential.

STEP 3

In [19]:
# Profit calculations

# Storing Key Values provided for calculation in variables
number_of_oil_wells=200
budget=100000000
revenue_per_barrel=4.5
revenue_per_unit_product=4500
risk_acceptance=0.025

# Calculating volume reserves sufficient for developing new wells without losses
cost_per_well=budget/number_of_oil_wells # Cost spent to develop each well

# cost_per_well=revenue_per_unit_product*volume -- for breaking even(money spent=money earned)
volume=cost_per_well/revenue_per_unit_product # Break- even volume
print()
print("Comparing break-even volume with the average volume of reserves in each region")
print(f"Break-even volume : {volume:.2f} thousand barrels")

print(f"Average predicted reserves for 'Region 0': {predictions_valid_0.mean():.2f} thousand barrels")
print(f"Average predicted reserves for 'Region 1': {predictions_valid_1.mean():.2f} thousand barrels")
print(f"Average predicted reserves for 'Region 2': {predictions_valid_2.mean():.2f} thousand barrels")



Comparing break-even volume with the average volume of reserves in each region
Break-even volume : 111.11 thousand barrels
Average predicted reserves for 'Region 0': 92.40 thousand barrels
Average predicted reserves for 'Region 1': 68.74 thousand barrels
Average predicted reserves for 'Region 2': 95.02 thousand barrels


Analysis: ALL three regions have average predicted reserves below the break-even volume of 111.11 thousand barrels, with region 1 being the lowest. This does not suggest that the regions are unprofitable. We need to now focus on selecting the top 200 wells as compared to random wells to proceed with our analysis. Our RMSE values of Region 0: 37.82, Region 1: 0.89, Region 2: 40.00 tell us that our confidence in profit calculation for Region 1 will be highest, and region 2 lowest. On the other hand, region 2 is closest to break-even which means we might have some top performing wells in region 2 but also might have some low performing wells, reducing the confidence level for this region. Selecting the top wells rather than average wells will give us the individual insight that we need to deepen our analysis, since mean value can get skewed because of some really low or really high performing wells in a region or by the number of wells.
This analysis confirms that our profit calculation function must focus on identifying and selecting the highest predicted wells from each region to make meaningful comparisons.

STEP 4

In [20]:
# Writing a function to calculate profit from top oil producing wells from a set of 500 randomly selected oil wells that were obtained in model predictions
import numpy as np

def calculate_profit(predicted_volumes,target_values,wells_count,budget,revenue_per_unit_product, sample_size=None):
    # Sampling logic
    if sample_size:
        sample_indices=np.random.choice(len(predicted_volumes), size=sample_size, replace=False)
        predicted_volumes=predicted_volumes[sample_indices]
        target_values= target_values.iloc[sample_indices]
        target_values= target_values.reset_index(drop=True)
        
    predictions_valid_series=pd.Series(predicted_volumes)
    top_wells_predictions=predictions_valid_series.nlargest(wells_count)
    target_values_series=pd.Series(target_values, index=predictions_valid_series.index)
    top_target_wells=target_values_series[top_wells_predictions.index]
    sum_top_wells_target=top_target_wells.sum()

    profit=(sum_top_wells_target*revenue_per_unit_product)-budget
    return profit

# Calculating Profit for all three Regions
profit_0=calculate_profit(predictions_valid_0,target_valid_0,200,100000000,4500, sample_size=500)
profit_1=calculate_profit(predictions_valid_1,target_valid_1,200,100000000,4500, sample_size=500)
profit_2=calculate_profit(predictions_valid_2,target_valid_2,200,100000000,4500, sample_size=500)

print(f"Predicted profits from top 200 oil producing wells in Regions 0,1 and 2 is:\n{profit_0:.2f}, {profit_1:.2f}, {profit_2:.2f} respectively")




Predicted profits from top 200 oil producing wells in Regions 0,1 and 2 is:
4725528.51, 6468076.40, 2709169.73 respectively


Analysis so far:

When we select top 200 wells - from 500 randomly chosen wells from each region, we get predicted profits as: $6.45 million for Region 0, $6.56 million for Region 1 and $2.84 million for region 2. This suggests that Region 1 has highest profits, followed by Region 0, with Region 2 having the lowest profit. All regions are profitable(positive values).
To get more confidence about our calculations, we will now use bootstrapping which will utilize sampling multiple times to get a distribution of profit.


STEP 5

In [21]:
# Using bootstrapping technique with 1000 samples to find distribution of profit

import numpy as np

state=np.random.RandomState(12345)

# Applying Bootstrapping Technique by defining a function

def bootstrapping(predictions_valid,target_valid):
    revenue_values=[]
    # Creating a validation dataset with target vaues of predicted set
    validation_data=pd.DataFrame({'predictions': predictions_valid, 'actual': target_valid}).reset_index(drop=True)

    # Creating a loop for creation of 1000 bootstrap samples
    for i in range(1000):
        # Selecting 500 random points for a sample
        validation_data_subsample=validation_data.sample(n=500,replace=True,random_state=i)
        #Extracting prediction and target subsamples from each bootstrap sample
        predictions_subsample=validation_data_subsample['predictions']
        target_subsample=validation_data_subsample['actual']
        # Calculating revenue by selecting best 200 points for profit calculation
        revenue_values.append(calculate_profit(predictions_subsample,target_subsample,200,100000000,4500))
    
    revenue_values= pd.Series(revenue_values)
    mean=revenue_values.mean()
    # Calculating lower and upper quantiles for risk assesment
    lower_quantile_region=revenue_values.quantile(q=0.025)
    higher_quantile_region=revenue_values.quantile(q=0.975)

    # Calculating risk of losses
    risk_of_loss=(revenue_values<0).mean()
    
    print(f"Mean profit: {mean:.2f}")
    print(f"Lower quantile of confidence interval: {lower_quantile_region:.2f}")
    print(f"Higher quantile of confidence interval:{higher_quantile_region:.2f}")
    print(f"Risk of losses: {risk_of_loss:.1%}\n")



# Displaying all Bootstrapping Analysis with 95% confidence interval values:

print("-"*50)
print("Region 0 Bootstrapping Analysis:")
print("-"*50)
bootstrapping(predictions_valid_0, target_valid_0)

print("-"*50)
print("Region 1 Bootstrapping Analysis:")
print("-"*50)
bootstrapping(predictions_valid_1, target_valid_1)

print("-"*50)
print("Region 2 Bootstrapping Analysis:")
print("-"*50)
bootstrapping(predictions_valid_2, target_valid_2)



--------------------------------------------------
Region 0 Bootstrapping Analysis:
--------------------------------------------------
Mean profit: 5875350.40
Lower quantile of confidence interval: -68632.67
Higher quantile of confidence interval:11980281.84
Risk of losses: 2.9%

--------------------------------------------------
Region 1 Bootstrapping Analysis:
--------------------------------------------------
Mean profit: 6777238.18
Lower quantile of confidence interval: 2170503.33
Higher quantile of confidence interval:12055145.31
Risk of losses: 0.4%

--------------------------------------------------
Region 2 Bootstrapping Analysis:
--------------------------------------------------
Mean profit: 5496130.57
Lower quantile of confidence interval: -750819.17
Higher quantile of confidence interval:11573189.61
Risk of losses: 4.6%



Results summary:

Region 0: 
    Mean Profit: $5,875,350
    95% Confidence Interval: ~ -$68,633 to ~$11,980,282
Region 1: 
    Mean Profit: $6,777,238
    95% Confidence Interval: ~ -$2,170,503 to ~$12,055,145
Region 2:
    Mean Profit: $5,496,131
    95% Confidence Interval: ~ -$750,819 to ~$11,573,189



Key Insights:
Region 1 comes out as a clear winner. This is because:
1. Highest Average Profit: $6.78M vs $5.88M(Region 0) and $5.50M(Region 2)
2. Lowest risk: The lower quantile is $2.17M, which means that there is less than 2.5% chance of loss
3. Most reliable prediction: RMSE was only 0.89 for Region 1
4. Risk assessment : Almost no risk of loss(lower bound is positive) vs Very low risk (only $68K potential loss for Region 0) and Moderate  risk ( upto 751k potential loss for Region 2)
5. Risk of losses: 0.4% as compared to 2.9% (Region 1) and 4.6% (Region 2)
                                                                                                                                                   
Hence, we can see that Region 1 is the best suited for development of oil wells.                                                                                                                                                   